File: sdk/src/chain.ts
import { Contract, JsonRpcProvider, Wallet } from "ethers";

/** Minimal ABI for ChainLog.sol — only the functions we call */
const CHAINLOG_ABI = [
  "function logAction(string agentId, bytes32 actionHash, string metadataURI) external",
  "function logBatch(string agentId, bytes32[] actionHashes, string[] metadataURIs) external",
  "function verifyAction(string agentId, uint256 index, bytes32 claimedHash) external view returns (bool)",
  "function getRecordCount(string agentId) external view returns (uint256)",
  "function totalActions() external view returns (uint256)",
  "event ActionLogged(string indexed agentId, bytes32 indexed actionHash, uint256 timestamp, address submitter)",
];

const BASE_SEPOLIA_RPC = "https://sepolia.base.org";

export interface ChainWriteResult {
  txHash: string;
}

/**
 * On-chain writer for ChainLog contract.
 */
export class ChainWriter {
  private contract: Contract;
  private wallet: Wallet;

  constructor(
    contractAddress: string,
    privateKey: string,
    rpcUrl: string = BASE_SEPOLIA_RPC
  ) {
    const provider = new JsonRpcProvider(rpcUrl);
    this.wallet = new Wallet(privateKey, provider);
    this.contract = new Contract(contractAddress, CHAINLOG_ABI, this.wallet);
  }

  /**
   * Write a single action to the blockchain.
   * Returns the transaction hash.
   */
  async logAction(
    agentId: string,
    actionHash: string,
    metadataURI: string = ""
  ): Promise<ChainWriteResult> {
    const tx = await this.contract.logAction(agentId, actionHash, metadataURI);
    const receipt = await tx.wait();
    return { txHash: receipt.hash };
  }

  /**
   * Write a single action without waiting for confirmation (fire-and-forget).
   * Returns the transaction hash immediately.
   */
  async logActionAsync(
    agentId: string,
    actionHash: string,
    metadataURI: string = ""
  ): Promise<string> {
    const tx = await this.contract.logAction(agentId, actionHash, metadataURI);
    return tx.hash;
  }

  /**
   * Write a batch of actions in a single transaction.
   */
  async logBatch(
    agentId: string,
    actionHashes: string[],
    metadataURIs: string[]
  ): Promise<ChainWriteResult> {
    const tx = await this.contract.logBatch(
      agentId,
      actionHashes,
      metadataURIs
    );
    const receipt = await tx.wait();
    return { txHash: receipt.hash };
  }

  /**
   * Verify an action hash against the on-chain record.
   */
  async verifyAction(
    agentId: string,
    index: number,
    claimedHash: string
  ): Promise<boolean> {
    return this.contract.verifyAction(agentId, index, claimedHash);
  }

  /**
   * Get the number of records for an agent.
   */
  async getRecordCount(agentId: string): Promise<number> {
    const count = await this.contract.getRecordCount(agentId);
    return Number(count);
  }

  /**
   * Get the signer's address.
   */
  get address(): string {
    return this.wallet.address;
  }
}

File: python/src/chainlog/chain.py
"""On-chain writer for ChainLog contract via web3.py."""

from __future__ import annotations

import logging

from web3 import AsyncWeb3, Web3
from web3.providers import AsyncHTTPProvider

logger = logging.getLogger(__name__)

CHAINLOG_ABI = [
    {
        "inputs": [
            {"name": "agentId", "type": "string"},
            {"name": "actionHash", "type": "bytes32"},
            {"name": "metadataURI", "type": "string"},
        ],
        "name": "logAction",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
    },
    {
        "inputs": [
            {"name": "agentId", "type": "string"},
            {"name": "actionHashes", "type": "bytes32[]"},
            {"name": "metadataURIs", "type": "string[]"},
        ],
        "name": "logBatch",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
    },
    {
        "inputs": [
            {"name": "agentId", "type": "string"},
            {"name": "index", "type": "uint256"},
            {"name": "claimedHash", "type": "bytes32"},
        ],
        "name": "verifyAction",
        "outputs": [{"name": "", "type": "bool"}],
        "stateMutability": "view",
        "type": "function",
    },
    {
        "inputs": [{"name": "agentId", "type": "string"}],
        "name": "getRecordCount",
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function",
    },
]

BASE_SEPOLIA_RPC = "https://sepolia.base.org"


class ChainWriter:
    """Async on-chain writer for ChainLog contract."""

    def __init__(
        self,
        contract_address: str,
        private_key: str,
        rpc_url: str | None = None,
    ) -> None:
        self._rpc_url = rpc_url or BASE_SEPOLIA_RPC
        self._w3 = AsyncWeb3(AsyncHTTPProvider(self._rpc_url))
        self._account = self._w3.eth.account.from_key(private_key)
        self._contract = self._w3.eth.contract(
            address=Web3.to_checksum_address(contract_address),
            abi=CHAINLOG_ABI,
        )

    async def log_action(
        self,
        agent_id: str,
        action_hash: str,
        metadata_uri: str = "",
    ) -> str:
        """Write a single action to the blockchain. Returns tx hash."""
        action_hash_bytes = bytes.fromhex(action_hash[2:])

        tx = await self._contract.functions.logAction(
            agent_id, action_hash_bytes, metadata_uri
        ).build_transaction(
            {
                "from": self._account.address,
                "nonce": await self._w3.eth.get_transaction_count(
                    self._account.address
                ),
            }
        )

        signed = self._account.sign_transaction(tx)
        tx_hash = await self._w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = await self._w3.eth.wait_for_transaction_receipt(tx_hash)
        return receipt["transactionHash"].hex()

    async def verify_action(
        self,
        agent_id: str,
        index: int,
        claimed_hash: str,
    ) -> bool:
        """Verify an action hash against the on-chain record."""
        claimed_bytes = bytes.fromhex(claimed_hash[2:])
        return await self._contract.functions.verifyAction(
            agent_id, index, claimed_bytes
        ).call()

    async def get_record_count(self, agent_id: str) -> int:
        """Get the number of records for an agent."""
        return await self._contract.functions.getRecordCount(agent_id).call()
